modules/BuildTools.Tasks/GenerateFileFromTemplate.cs

```diff
@@ -0,0 +1,179 @@
+                return false;
+            }
+
+            var values = MSBuildListSplitter.GetNamedProperties(Properties);
+            var template = File.ReadAllText(TemplateFile);
+
+            var result = Replace(template, values);
+            Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
+            File.WriteAllText(outputPath, result);
+
+            FileWrites = Path.GetFullPath(outputPath);
+
+            return true;
+        }
+
+        internal string Replace(string template, IDictionary<string, string> values)
+        {
+            var sb = new StringBuilder();
+            var line = 1;
+            for (var i = 0; i < template.Length; i++)
+            {
+                var ch = template[i];
+                if (ch == '\n')
+                {
+                    line++;
+                }
+
+                if (ch == '`')
+                {
+                    i++;
+                    if (i >= template.Length)
+                    {
+                        // slash ends doc
+                        sb.Append(ch);
+                        break;
+                    }
+
+                    ch = template[i];
+                    if (ch != '$' && ch != '`')
+                    {
+                        // Not a known escape character
+                        sb.Append('`');
+                    }
+
+                    sb.Append(ch);
+                    continue;
+                }
+
+                if (ch != '$')
+                {
+                    sb.Append(ch);
+                    continue;
+                }
+
+                i++;
+                if (i >= template.Length)
+                {
+                    sb.Append('$');
+                    break;
+                }
+
+                if (template[i] != '{')
+                {
+                    sb.Append('$').Append(template[i]);
+                    continue;
+                }
+
+                var varNameSb = new StringBuilder();
+                i++;
+                for (; i < template.Length; i++)
+                {
+                    var nextCh = template[i];
+                    if (nextCh != '}')
+                    {
+                        varNameSb.Append(nextCh);
+                    }
+                    else
+                    {
+                        var varName = varNameSb.ToString();
+                        if (values.TryGetValue(varName, out var value))
+                        {
+                            sb.Append(value);
+                        }
+                        else
+                        {
+                            Log.LogWarning(null, null, null, TemplateFile,
+                                line, 0, 0, 0,
+                                message: "No property value is available for '{0}'",
+                                messageArgs: new[] { varName });
+                        }
+
+                        varNameSb.Clear();
```

Comment:
No need to clear if you aren't reusing the SB? For perf you should reuse but I'm not sure how much of a concern it is here.